### ISSUE TITLE
Automatic gradient accumulation for batch size tuning

### ISSUE THREAD

**Issue by bilelomrani1** [Lightning-AI/lightning#17393]
### Description & Motivation

When training different model sizes on a different number of devices or different hardware, the batch size needs to be carefully tuned in order to achieve maximum GPU utilization without incurring Out Of Memory Error. Lightning currently has a `BatchSizeFinder` tuner, but it does not work with distributed training, which is precisely the setup where tuning the batch size is a tricky and expensive process.

### Pitch

One appealing solution could be to use an automatic microbatching strategy with gradient accumulation. In terms of user API it could be used with

```python
Trainer(
    ...
    accumulate_grad_batches="auto",
    ...
)
```
The strategy is the following: we attempt to train on the full batch size provided by the user, catch any Cuda OOM exception, half the batch size if needed and retry training until the forward step succeeds. These micro batches are then accumulated to simulate the effective batch size.

(Note that this does mean that we should be able to slice through the batches provided by the user which can be challenging if the user uses custom data structures or layout, but should be doable in the most standard practice)

This feature maximizes efficiency while making it easier for the user to scale in and out without having to constantly tuning the batch size.

### Alternatives

If you have any recommendation or alternatives for automating the batch size in distributed training, I would be very thankful for that!

### Additional context

I came across this feature when reading through the documentation of the fantastic Mosaic ML library [Automatic Microbatching](https://docs.mosaicml.com/projects/composer/en/latest/notes/auto_microbatching.html) and found the idea very cool and useful. It would be a very nice addition to Lightning as there are no way of automatically tuning the batch size in distributed settings afaik.
[Their implementation](https://github.com/mosaicml/composer/blob/dev/composer/trainer/trainer.py) is open source, under Apache 2.0.

**Comment by FarzanT** [comment]
I think that's a great idea!
Additionally, it would be nice if Lightning could automatically scale batch sizes whenever it encounters OOM errors. 
E.g. when training models with variable length input, its difficult to calculate batch size for the largest length input. Even with doing so, you'd be forced with setting a batch size that may be too small for shorter length sequences, wasting GPU resources.
Any pointers from maintainers on how to implement this would be appreciated!

**Comment by FarzanT** [comment]
Thanks @bilelomrani1!

@awaelchli I think this feature is quite useful for variable-length data across NLP, vision, computational biology, etc. Do you have any pointers on how to bring such a functionality to Lightning? There might be already some overlap with the [`batch_size_finder`](https://github.com/Lightning-AI/lightning/blob/ca30fd7752582201a3966806c92e3acbbaf2a045/src/lightning/pytorch/callbacks/batch_size_finder.py).

Thank you all!

**Comment by IanMaquignaz** [comment]
Definitely need this! I have manually implemented this logic with [Toma](https://github.com/BlackHC/toma) and it works... but it would be great to have it nicely packaged. 

My use case is SLURM on HPC platforms. It was previously taking me hours to manually tune batch size per specific hardware capacities. 

If you really want to overpower this feature: Clusters often have sets of nodes with different hardware configurations, but it's generally a good idea to only use homogeneous nodes. It would be absolutely amazing to have gradient accumulation balanced on a per-node-basis, allowing for non-homogeneous nodes and full-cluster utilization.

**Comment by FarzanT** [comment]
@IanMaquignaz If your code can somehow be transferred to Lightning that'll be awesome! I'm sure it'll greatly help whoever uses variable-length models!

**Comment by IanMaquignaz** [comment]
Unfortunately my code is unstable. It doesn't work on multiple nodes and often leaves a memory footprint on the GPU which causes OOM errors. 

That being said, the logic is quite simple:
```
from toma import toma

@toma.batch(initial_batchsize=config['batch_size'])
def train_with_toma(batchsize):
    # Adjust batchsize
    config.batch_size = batchsize
    if config.batch_size_effective < config.batch_size:
        config.batch_size = config.batch_size_effective
    if config.accumulated_batches is None \
        or config.batch_size * config.accumulated_batches != config.batch_size_effective:
        config.accumulated_batches = config.batch_size_effective // config.batch_size
        config.accumulated_batches = max(1, config.accumulated_batches) 
        print(
            f"Adjusted Batch Size!" \
            + f"\n\t Original batch_size={config_base['batch_size']}" \
            + f"\n\t New batch_size={config.batch_size}" \
            + f"\n\t New accumulated_batches={config.accumulated_batches}"
        )

     # Load your model
     # Create your trainer
     # Fit your model
```

**Comment by SuperSecureHuman** [comment]
Would love to see updates on this..

My senario:

I would like to have my batch_size locked in, and let lightning figure out whats the best match of grad_accum along with per_device_batch_size so that I can maximize the throughput of my slurm jobs.
